File: Array/Python/_1572_MatrixDiagonalSum.py
from typing import List


class Solution:
    # Iteration
    # Time Complexity: BigO(N)
    # Space Complexity: BigO(1)
    def diagonalSum(self, mat: List[List[int]]) -> int:
        sum = 0
        rt = len(mat) - 1

        for row in range(0, len(mat)):
            sum += mat[row][row]
            if row != rt:
                sum += mat[row][rt]
            rt -= 1

        return sum


ins = Solution()

print(ins.diagonalSum([[1, 2, 3], [4, 5, 6], [7, 8, 9]]))
print(ins.diagonalSum([[1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1]]))
print(ins.diagonalSum([[5]]))

File: Array/TypeScript/_1572_MatrixDiagonalSum.ts
/**
 * Iteration
 * Time Complexity: BigO(N)
 * Space Complexity: BigO(1)
 */
function diagonalSum(mat: number[][]): number {
  let sum = 0;
  let rt = mat.length - 1;

  for (let row = 0; row < mat.length; row++) {
    sum += mat[row][row];
    if (row != rt) {
      sum += mat[row][rt];
    }
    rt--;
  }

  return sum;
}

console.log(
  diagonalSum([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
  ])
);
console.log(
  diagonalSum([
    [1, 1, 1, 1],
    [1, 1, 1, 1],
    [1, 1, 1, 1],
    [1, 1, 1, 1],
  ])
);
console.log(diagonalSum([[5]]));
